daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala

```diff
@@ -24,10 +24,49 @@ import scala.util.hashing.MurmurHash3
   */
 sealed trait SValue {
 
-  import SValue._
+  import SValue.{SValue => _, _}
 
-  def toValue: V[V.ContractId] =
+  def toValue: V[V.ContractId] = SValue.toValue(this)
+
+  def mapContractId(f: V.ContractId => V.ContractId): SValue =
     this match {
+      case SContractId(coid) => SContractId(f(coid))
+      case SEnum(_, _, _) | _: SPrimLit | SToken | STNat(_) | STypeRep(_) => this
+      case SPAP(prim, args, arity) =>
+        val prim2 = prim match {
+          case PClosure(label, expr, vars) =>
+            PClosure(label, expr, vars.map(_.mapContractId(f)))
+          case other => other
+        }
+        val args2 = mapArrayList(args, _.mapContractId(f))
+        SPAP(prim2, args2, arity)
+      case SRecord(tycon, fields, values) =>
+        SRecord(tycon, fields, mapArrayList(values, v => v.mapContractId(f)))
+      case SStruct(fields, values) =>
+        SStruct(fields, mapArrayList(values, v => v.mapContractId(f)))
+      case SVariant(tycon, variant, rank, value) =>
+        SVariant(tycon, variant, rank, value.mapContractId(f))
+      case SList(lst) =>
+        SList(lst.map(_.mapContractId(f)))
+      case SOptional(mbV) =>
+        SOptional(mbV.map(_.mapContractId(f)))
+      case SMap(isTextMap, value) =>
+        SMap(
+          isTextMap,
+          value.iterator.map { case (k, v) => k.mapContractId(f) -> v.mapContractId(f) },
+        )
+      case SAny(ty, value) =>
+        SAny(ty, value.mapContractId(f))
+    }
+}
+
+object SValue {
+
+  def toValue(v: SValue, maxNesting: Int = V.MAXIMUM_NESTING): V[V.ContractId] = {
```

Comment:
should we have some unit tests for this?

Reply:
added

Reply:
9e21da583838c14d6581a950b15e9d10386543da